File: scripts/integrations/deploy_integration.py
#!/usr/bin/env python3
"""
deploy_integration.py

PURPOSE: Deploy built integration to Prismatic platform

USAGE: python deploy_integration.py <project-directory>

PARAMETERS:
  project-directory - Path to integration project directory

EXIT CODES:
  0 - Success: Integration deployed
  1 - Error: Project directory not found or not built
  2 - Error: Authentication failed
  3 - Error: Import failed

"""

import os
import subprocess
import sys
import time

SHARED_DIR = os.path.join(os.path.dirname(__file__), '..', 'shared')
sys.path.insert(0, SHARED_DIR)

from graphql import ensure_authenticated, GraphQLError
from prism_retry import run_prism_mutation


def deploy_integration(project_dir):
    """Deploy integration to Prismatic."""
    if not os.path.isdir(project_dir):
        print(f"❌ Project directory not found: {project_dir}")
        return 1

    dist_dir = os.path.join(project_dir, "dist")
    if not os.path.isdir(dist_dir):
        print("❌ Build artifacts not found")
        print("")
        print("You need to build the integration first.")
        print(f"Run: python build_integration.py {project_dir}")
        return 1

    print(f"🚀 Deploying integration: {project_dir}")
    print("")

    # Load authentication credentials (token and URL)
    try:
        ensure_authenticated()
    except GraphQLError as e:
        print(f"❌ {e}")
        return 2

    cmd = ["prism", "integrations:import"]
    print(f"Running: {' '.join(cmd)}")
    print("")

    try:
        result = run_prism_mutation(
            cmd,
            cwd=project_dir,
            timeout=60,
        )

        if result.returncode == 0:
            print("✅ Integration deployed successfully!")
            print("")

            if result.stdout:
                print(result.stdout)

            # Wait for integration to be fully available
            print("")
            print("⏳ Waiting 5 seconds for integration to be fully available...")
            time.sleep(5)
            print("")

            print("📋 Next steps:")
            print("  - View and configure integration in Prismatic web app")
            print(
                f"  - Test: python scripts/test_integration.py <integration-id> --integration-dir {project_dir}"
            )
            print(
                f"  - Package for download: python scripts/package_for_download.py {project_dir}"
            )
            print("")
            print("⚠️  TESTING: Always use scripts/test_integration.py")
            print("   Do NOT use prism commands directly for testing")

            return 0
        else:
            print("❌ Deployment failed")
            print("")

            if result.stderr:
                print("Error output:")
                print(result.stderr)

            if result.stdout:
                print("")
                print("Standard output:")
                print(result.stdout)

            print("")
            print("💡 Troubleshooting:")
            print("  - Ensure you're authenticated: python check_prism_access.py")
            print("  - Verify build succeeded: check dist/ directory")
            print("  - Check for validation errors in your integration definition")
            return 3

    except FileNotFoundError:
        print("❌ Prism CLI not found")
        print("Run setup_prerequisites.py first")
        return 3

    except subprocess.TimeoutExpired:
        print("❌ Deployment timed out (60 seconds)")
        print("")
        print("The deployment took longer than expected.")
        print("This could indicate network issues or a large integration.")
        return 3

    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        print("")
        print(f"Error type: {type(e).__name__}")
        print(f"Error message: {str(e)}")
        return 3


def main():
    if len(sys.argv) < 2:
        print("❌ No project directory provided")
        print("Usage: python deploy_integration.py <project-directory>")
        return 1

    project_dir = sys.argv[1]
    return deploy_integration(project_dir)


if __name__ == "__main__":
    sys.exit(main())

File: scripts/integrations/deploy-integration.ts
#!/usr/bin/env npx tsx
/**
 * deploy-integration.ts
 *
 * PURPOSE: Deploy built integration to the platform
 *
 * USAGE: npx tsx deploy-integration.ts <project-directory>
 *
 * EXIT CODES:
 *   0 - Success: Integration deployed
 *   1 - Error: Project directory not found or not built
 *   2 - Error: Authentication failed
 *   3 - Error: Import failed
 */

import { existsSync } from "node:fs";
import { ensureAuthenticated, GraphQLError } from "../shared/graphql.js";
import { runPrismMutation } from "../shared/prism-retry.js";

function sleepSync(ms: number): void {
  const end = Date.now() + ms;
  while (Date.now() < end) {
    // busy wait — acceptable for short CLI delays
  }
}

function deployIntegration(projectDir: string): number {
  if (!existsSync(projectDir)) {
    console.log(`Project directory not found: ${projectDir}`);
    return 1;
  }

  const distDir = `${projectDir}/dist`;
  if (!existsSync(distDir)) {
    console.log("Build artifacts not found");
    console.log("");
    console.log("You need to build the integration first.");
    console.log(`Run: npx tsx scripts/integrations/build-integration.ts ${projectDir}`);
    return 1;
  }

  console.log(`Deploying integration: ${projectDir}`);
  console.log("");

  try {
    ensureAuthenticated();
  } catch (e) {
    if (e instanceof GraphQLError) {
      console.log(e.message);
      return 2;
    }
    throw e;
  }

  const cmd = ["prism", "integrations:import"];
  console.log(`Running: ${cmd.join(" ")}`);
  console.log("");

  try {
    const result = runPrismMutation(cmd, { timeout: 60, cwd: projectDir });

    if (result.returncode === 0) {
      console.log("Integration deployed successfully!");
      console.log("");

      if (result.stdout) console.log(result.stdout);

      console.log("");
      console.log("Waiting 5 seconds for integration to be fully available...");
      sleepSync(5000);
      console.log("");

      console.log("Next steps:");
      console.log("  - View and configure integration in the web app");
      console.log(
        `  - Test: npx tsx scripts/shared/test-integration.ts <integration-id> --integration-dir ${projectDir}`
      );
      console.log(
        `  - Package for download: npx tsx scripts/shared/package-for-download.ts ${projectDir}`
      );
      console.log("");
      console.log("TESTING: Always use scripts/test-integration.ts");
      console.log("   Do NOT use prism commands directly for testing");

      return 0;
    } else {
      console.log("Deployment failed");
      console.log("");

      if (result.stderr) {
        console.log("Error output:");
        console.log(result.stderr);
      }
      if (result.stdout) {
        console.log("");
        console.log("Standard output:");
        console.log(result.stdout);
      }

      console.log("");
      console.log("Troubleshooting:");
      console.log("  - Ensure you're authenticated: npx tsx scripts/shared/check-prism-access.ts");
      console.log("  - Verify build succeeded: check dist/ directory");
      console.log("  - Check for validation errors in your integration definition");
      return 3;
    }
  } catch (e: unknown) {
    if (e instanceof Error && e.message.includes("ENOENT")) {
      console.log("Prism CLI not found");
      console.log("Run prerequisites first");
      return 3;
    }
    console.log(`Unexpected error: ${e}`);
    return 3;
  }
}

function main(): number {
  if (process.argv.length < 3) {
    console.log("No project directory provided");
    console.log("Usage: npx tsx deploy-integration.ts <project-directory>");
    return 1;
  }

  return deployIntegration(process.argv[2]);
}

process.exit(main());
